First milestone of keysafe network

yyd106 opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/947 < please fill this in with the PR number of your application.

alxs commented on deliveries/keysafe_network-milestone_1.md:
```diff
@@ -0,0 +1,23 @@
+
+**The [invoice form :pencil:](https://forms.gle/8Wx7nxtq8fKrsuEz8) has been filled out correctly for this milestone and the delivery is according to the official [milestone delivery guidelines](https://github.com/w3f/General-Grants-Program/blob/master/grants/milestone-deliverables-guidelines.md).** 
+
+* **PR Link:** [Keysafe](https://github.com/w3f/Grants-Program/pull/947). 
```
```suggestion
* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/keysafe_network.md
```

alxs commented:
Thanks for the delivery @yyd106. We'll look into it as soon as possible.

0xCaso commented:
Hi @yyd106! I'm beginning the evaluation, I'll reach out if I need your help.

0xCaso commented:
Here I am, I'm trying to run the tests in the docker instance, but I got these two errors:
```
error[E0425]: cannot find value `db_port` in this scope
   --> src/main.rs:106:31
    |
106 |         db_user, db_password, db_port, db_name);
    |                               ^^^^^^^ not found in this scope

error[E0425]: cannot find value `db_name` in this scope
   --> src/main.rs:106:40
    |
106 |         db_user, db_password, db_port, db_name);
    |                                        ^^^^^^^ not found in this scope
```
Note: I had to add `ENV CARGO_NET_GIT_FETCH_WITH_CLI=true` in the Dockerfile

yyd106 commented:
@0xCaso ,thanks. @jarodzz will support here.

jarodzz commented:
There was a mistake in previous version, the code didn't fit the new config. Now it is fixed. Please pull the latest version and try again. Let me know if there's any issues.

0xCaso commented:
Thanks @jarodzz. Now I've got this error, at the end of the build:
```
error: linking with `cc` failed: exit status: 1
```
Also, I'm running it with a Mac M1, so don't know if it could be a problem also for the docker

jarodzz commented:
@0xCaso I tested the process on a ubuntu machine, it worked. I am testing it on a mac machine now, not M1 though, hopefully I can reproduce the problem. I'll let you know.

0xCaso commented:
Other than this error, I have a question about the whole solution security.
Is it actually *safe* sending to a website, so online, a private key? I've seen you've talked about security issues [here](https://github.com/keysafe-protocol/documents/blob/main/README.md#the-attacks-and-security-models), but I was asking me that question too.

yyd106 commented:
> Other than this error, I have a question about the whole solution security. Is it actually _safe_ sending to a website, so online, a private key? I've seen you've talked about security issues [here](https://github.com/keysafe-protocol/documents/blob/main/README.md#the-attacks-and-security-models), but I was asking me that question too.

@0xCaso 
In the current version, the user key is split into three pieces, two of which are stored inside the TEE on two different Keysafe nodes and be protected by DAuth. 

In the next version, we are going to make the solution self-custody. We will involve Face scan + iCloud/Keychain, to protect the 3rd key share in iCloud. In this version, 3 keys are in no one can get the original key of the user. The three key shares are respectively stores on the user side, the keysafe node TEE, as well. icloud/keychain, no one can recover the full key.

0xCaso commented:
Thanks a lot @yyd106 for the explanation, definitely it makes sense then!
Are there any news for the mac testing?

jarodzz commented:
@0xCaso Hi, please pull the latest code and try again. It was a bug in makefile, fixed in the latest version.

0xCaso commented:
Hi @jarodzz! Thanks for the update. I managed to build the package, and then also to run the tests. I couldn't run the tests before building the package, so I'll point this out in the evaluation (to invert the order of the two commands).

0xCaso commented:
Also @jarodzz, is it possible to run the app in the docker, with these commands (after the build with `make SGX_MODE=SW`)
```
cd bin
ln -s ../certs .
./app
```
or are they just for the setup without docker?
In case it's possible, I'm running into this error:
```
thread 'main' panicked at 'called `Result::unwrap()` on an 
`Err` value: No such file or directory (os error 2)', src/main.rs:147:57
```

0xCaso commented:
Also another question, the polkadot.js integration isn't in the demo dapp, right? If this is the case, would it be too long to integrate it?

0xCaso commented:
> > Also another question, the polkadot.js integration isn't in the demo dapp, right? If this is the case, would it be too long to integrate it?
> 
> @0xCaso The demo link in the user guide is an old one, It's a mistake, I have fixed with the latest demo link: https://keysafe-front-keysafe.vercel.app/#/
> 
> The initial integration timing was two months. However, in the development process, we found that the core technology DAuth is more complex than imagined, because it involves the transformation of Https protocol and SMTP protocol in TEE. So we spent about three months for the technical validation. Therefore, we postponed the delivery time accordingly.

Great, thanks a lot also for the clarification. Also, does the Polkadot account importing work? Because, inserting the mnemonics, it can't correctly take the account.

yyd106 commented:
> Great, thanks a lot also for the clarification. Also, does the Polkadot account importing work? Because, inserting the mnemonics, it can't correctly take the account.

@0xCaso At this stage, we have not introduced MPC threshold signature, instead we are now using Shamir Secret Sharing. So it works for mnemonics as well. In the future, we will transition to the MPC+TSS scheme, and then we need to develop the polkadot MPC key generation method.

@jarodzz is looking into the #639

0xCaso commented:
> > Great, thanks a lot also for the clarification. Also, does the Polkadot account importing work? Because, inserting the mnemonics, it can't correctly take the account.
> 
> @0xCaso At this stage, we have not introduced MPC threshold signature, instead we are now using Shamir Secret Sharing. So it works for mnemonics as well. In the future, we will transition to the MPC+TSS scheme, and then we need to develop the polkadot MPC key generation method.
> 
> @jarodzz is looking into the comment

Thanks a lot for the clarification @yyd106. The integration with the Polkadot accounts is one of the deliverables of this milestones. Do you want to amend it for the next milestone, if it takes too much for the integration, or do you prefer keeping it for this one and update me when it's finished?

yyd106 commented:
> > > Great, thanks a lot also for the clarification. Also, does the Polkadot account importing work? Because, inserting the mnemonics, it can't correctly take the account.
> > 
> > 
> > @0xCaso At this stage, we have not introduced MPC threshold signature, instead we are now using Shamir Secret Sharing. So it works for mnemonics as well. In the future, we will transition to the MPC+TSS scheme, and then we need to develop the polkadot MPC key generation method.
> > @jarodzz is looking into the comment
> 
> Thanks a lot for the clarification @yyd106. The integration with the Polkadot accounts is one of the deliverables of this milestones. Do you want to amend it for the next milestone, if it takes too much for the integration, or do you prefer keeping it for this one and update me when it's finished?

@0xCaso The Polkadot account support is already worked here. As you can see ![](https://github.com/keysafe-protocol/documents/blob/main/pictures/polkadot_integration.png?raw=true)

yyd106 commented:
> @yyd106 I'm sorry, when I entered the mnemonic and submitted to save the account I had this error: <img alt="Screenshot 2022-12-05 at 18 20 45" width="584" src="https://user-images.githubusercontent.com/22419450/205951316-2333c325-0db6-47d8-98d6-6918fe5eb9f9.png"> After this, I haven't been redirected to the home page, and I didn't see the Polkadot account saved. Refreshing the page, it showed. Thanks.
> 
> **Note**: this happens with every network, but I didn't notice it since in the last demo the error didn't appear, and in the new one I was trying only the Polkadot account

@0xCaso We can't reproduce from our side, we are looking into it. You are using this link right? https://keysafe-front-keysafe.vercel.app/

0xCaso commented:
@yyd106 It's weird, now everything works fine also for me. By the way, as soon as I'll receive the response to the https://github.com/w3f/Grant-Milestone-Delivery/pull/639#issuecomment-1337573195 I'll post my evaluation

0xCaso reviewed (changes requested):
@yyd106 It's weird, now everything works fine also for me. By the way, as soon as I'll receive the response to the https://github.com/w3f/Grant-Milestone-Delivery/pull/639#issuecomment-1337573195 I'll post my evaluation

0xCaso commented:
@yyd106 @jarodzz any update?

yyd106 commented:
> @yyd106 @jarodzz any update?

We are fixing the Email sent function issue inside SGX in Docker, could we take few days more from now? Definitely will let you know once we are done

0xCaso commented:
Thanks for the response, no problem!

0xCaso commented:
Just to let you know, I will be on vacation till Jan 3, so if I won't see any update when I'm back I'll ask for termination. In the meantime, merry xmas and happy new year :)

yyd106 commented:
> Just to let you know, I will be on vacation till Jan 3, so if I won't see any update when I'm back I'll ask for termination. In the meantime, merry xmas and happy new year :)
Thank you for the reminding. We will update before Jan 3. Merry xmas!

yyd106 commented:
> Just to let you know, I will be on vacation till Jan 3, so if I won't see any update when I'm back I'll ask for termination. In the meantime, merry xmas and happy new year :)

Hi @0xCaso we have fixed those mentioned requirements and updated the [test doc](https://github.com/keysafe-protocol/keysafe-app/blob/polkadot/README.md)

0xCaso commented:
Thanks for the update! I'm trying to run the docker, but the command `make -f MakeHwFile SGX_MODE=SW` gives me this error:
```
ar rcsD lib/libEnclave_u.a app/Enclave_KS_u.o
ar: lib/libEnclave_u.a: No such file or directory
MakeHwFile:133: recipe for target 'lib/libEnclave_u.a' failed
make: *** [lib/libEnclave_u.a] Error 1
```
Maybe the `MakeHwFile` vars are misconfigured?
Could it be for [this](https://github.com/keysafe-protocol/keysafe-app/blob/polkadot/MakeHwFile#L28)?

ksrust commented:
Hi, @0xCaso.
To generate lib/libEnclave_u.a, we need app/Enclave_KS_u.o. This file comes from the enclave part, which should be generated after you execute "bash /root/get_ks_sgx.sh".
You can try this to re-generate it.
$ cd /root/
$ bash /root/get_ks_sgx.sh 
And go back to the make step: 
$ cd /root/incubator-teaclave-sgx-sdk/samplecode/keysafe-app/;
$ make -f MakeHwFile SGX_MODE=SW

0xCaso commented:
> Hi, @0xCaso. To generate lib/libEnclave_u.a, we need app/Enclave_KS_u.o. This file comes from the enclave part, which should be generated after you execute "bash /root/get_ks_sgx.sh". You can try this to re-generate it. $ cd /root/ $ bash /root/get_ks_sgx.sh And go back to the make step: $ cd /root/incubator-teaclave-sgx-sdk/samplecode/keysafe-app/; $ make -f MakeHwFile SGX_MODE=SW

I already did it, as you described these steps in the README.

jarodzz commented:
Hi, @0xCaso .
My mistake. There was a tmp dir that was created in my environment, but not in the MakeHwFile. I just updated the code.
If your docker image is ready, you can pull the latest code outside docker, and make again inside docker.

0xCaso commented:
Thanks a lot @jarodzz, now the app starts!

Now, in the last step of the README, I have to run the front-end. In this line:
```
docker run --rm -p 3000:3000 -e REACT_APP_BASE_URL='https://<your-ip-address>:30000/ks' keysafe-frontend
```
- is `30000` the right port?
- do I have to put my IP address instead of `<your-ip-address>`, or is `localhost` good? For the first option, I don't think my IP would work as I'm running the app in a docker, which is in a remote machine (not my local one).

0xCaso reviewed (commented):
Thanks a lot again, with `http` it worked (I invite to edit the README for this then);

Now, I'm setting the Auth Conditions, but if I select `gauth` the applications breaks. This is the log:
```
middleware is verifying token.
getting path /ks/register_gauth
analyze header eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJtYXR0ZW9Ad2ViMy5mb3VuZGF0aW9uIiwiZXhwIjoxNjczNDUyNDUwfQ.zNw8QJZfKImeejUY82xYj6W8agNFBX46RNCxG7bcSsU with keysafe_password_for_test
path that requires token and verifying.
calling gen gauth secret
Illegal instruction (core dumped)
```
`email` and `password` methods work though.

Also, I've found these logs inside `/keysafe-app/bin/logs/ks.log`, but I don't know if it's related to that error:
```
2023-01-04T16:02:41.467816612+00:00 - Error registering
2023-01-04T16:02:41.488942770+00:00 - Starting 16 workers
2023-01-04T16:02:41.488966765+00:00 - Actix runtime found; starting in Actix runtime
2023-01-04T16:02:41.489466038+00:00 - Specified path is not a directory: "./public"
```

jarodzz commented:
@0xCaso The 'Illegal instruction (core dumped)' comes from sgx/tee, which is might be environment related. Are you testing on a ubuntu server or mac? And let me know your CPU type. 
I tested on a ubuntu 22 with an intel CPU, it's working fine. 
I'll start over again to try to reproduce this issue.

0xCaso commented:
@jarodzz I'm using Ubuntu 20.04, and this is some data of the CPU:
```
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Model name:                      AMD Ryzen 9 5950X 16-Core Processor
```

0xCaso reviewed (commented):
@jarodzz I'm using Ubuntu 20.04, and this is some data of the CPU:
```
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Model name:                      AMD Ryzen 9 5950X 16-Core Processor
```

jarodzz commented:
Hi, @0xCaso 

According to intel sgx community, AMD Ryzen does not support SGX simulation mode.
https://community.intel.com/t5/Intel-Software-Guard-Extensions/Build-run-issue-of-inte-sgx-ssl-in-SIM-mode-on-AMD-Ryzen-5/td-p/1233407

Is it possible to find a machine with Intel CPU?

0xCaso reviewed (commented):
I tried with another laptop with the Intel CPU, and everything worked, thanks a lot!

I'll publish the evaluation soon :)

0xCaso commented:
[Here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/keysafe_network_1_0xCaso.md) you can find the evaluation. Let me know if you have any questions!

0xCaso reviewed (changes requested):
[Here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/keysafe_network_1_0xCaso.md) you can find the evaluation. Let me know if you have any questions!

jarodzz commented:
@0xCaso Glad it finally worked. 
Thank you for your support. Your evaluation report is clear and comprehensive, all the issues are valuable to keysafe for future development.
As for the License and ink! unit test, I'll start working on these 2 urgent fixes right now, will get back to you in a day or two.
As for other issues, I'll discuss with the team and update our comments later.

jarodzz commented:
Hi, @0xCaso.
About the 2 urgent issues:
1. I have updated all license to Apache 2.0.
2. I'm working on ink! smart contract, adding more unit tests. When trying to build and test keysafe smart contract, I find out that the ink! library is updated, along with some grammar change, so I have to make some minor changes to adapt to this new version. Will update status in the next couple days.

0xCaso commented:
Thanks @jarodzz for letting me know, looking forward the update!

0xCaso reviewed (changes requested):
Thanks @jarodzz for the update, and I'm also happy to read that!

By the way, before accepting the delivery:
- I'd like to read your answers (and fix the minor issues, like in the README of the `keysafe-app` repo) to the comments I wrote in the [evaluation](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/keysafe_network_1_0xCaso.md#other-notes);
- It would be great if in the `contracts` repo you could provide the environment to test the ink contract and a minimal guide in a README.

jarodzz commented:
Hi, @0xCaso.

I  updated Readme in keysafe-app, fixed 2 issues that you mentioned.
And I created a new Readme in contracts/ink/ with the steps of how to build and test the contracts.

I'll work on other issues as well. Get back to you soon.

0xCaso commented:
Thanks a lot @jarodzz and @yyd106 for the changing and comments!

All clear, then I'm looking forward your next update!

yyd106 commented:
@0xCaso We have fixed all the issue in the evaluation. Some of them are explainable
### 0.c Testiing Guide
* Which functions have been covered? Runnable tests cover just the authentication functions, or am I wrong?
The decentralized authentication functions is the main part of our technology, the TEE part is hard to implement since it requires some sort of tools and need to setup simulation env. Can we implement these settings and tools into next milestore? Since next milestone will have more complicated flows inside TEE and current is quite simple.
* How high is the coverage?
The unitest covered most of the logic except code inside TEE.
### Other notes
* Is the transfer functionality working? If yes, I couldn't manage to transfer funds (cannot click "confirm" button);
The confirm button will be available once user input all necessary info;
* Balance is wrong (I've tried with Boba Rinkeby, and it doesn't show the correct balance);
Fixed
* I can't understand what's the point of choosing the Network for different EVMs blockchains (if I put a private key, it should be the same for every Network);
In this demo version, we simplified the network options, only ETH testnet, Polkdot is supported. A more user-friendly multi-chain support design will be seen in the [Orcas wallet](https://orcas.gitbook.io/orcas-wallet/);
* When I try to recover an account's keys, it asks the Network and the Key: I shouldn't be able to export an Ethereum Key with "Polkadot" selected, but I can; maybe it has more sense to ask only the Key;
This has been fixed, only Chain A's accounts could be selected when Chain A is selected;
* When I set the same signup email for the authentication, it shouldn't re-ask the verification;
Since the signup flow and sign-in flow is basically the same, so we merged the two process in the demo;
* GitHub authentication redirects to the old frontend, so it's not possible to use it for the recovery; also, Google and Twitter are not implemented yet in My Web2 Accounts, right?
This is due to a DNC setting problem, it has been fixed;
* I assumed My Auth Conditions and General Settings are not yet implemented;
Correct, will implement it in the official product, the Orcas Wallet will implement the "change email" function, with is similar with "My Auth Conditions" setting;
* In the old UI, during the signup phase, if I switched through Google Authenticator and password while setting the recovery methods, the password was filled with strange characters, somehow seemed related to the generated QR code (couldn't retry in the new UI, maybe try to check if the error persists);
The old UI is abandoned and with the fix of DNS, it will not be jumped from the new UI;
* Generally UI/UX is to improve, some things are not so clear: for example how many authentications are needed in order to retrieve the keys; but it's pretty normal I guess since this is a demo.
Yes the UI is crude. We will deliver Orcas Wallet with a great UI/UE.

0xCaso commented:
Thanks @yyd106 and @jarodzz for the changes and for the answers/comments. I'm happy to approve this and ask the same to my team.

0xCaso reviewed (approved):
Thanks @yyd106 and @jarodzz for the changes and for the answers/comments. I'm happy to approve this and ask the same to my team.

0xCaso commented:
One last thing @yyd106: I'd kindly ask you to re-send the invoice for this milestone with the following additions:
- the W3F VAT ID (which you can find in the [invoice form](https://docs.google.com/forms/d/e/1FAIpQLSfmNYaoCgrxyhzgoKQ0ynQvnNRoTmgApz9NrMp-hd8mhIiO0A/viewform));
- the VAT amount (0).

Also, recently we changed our guidelines, and I would ask you if it's possible to change USDT on Ethereum as payment method to another option choosable from [here](https://github.com/w3f/Grants-Program#pencil-process) (so if you want to stay with Ethereum you should change it to USDC or DAI).

If yes, then please update the currency also in the invoice.

yyd106 commented:
@0xCaso Thanks a lot for the support. I have re-uploaded the invoice form according to the suggestion. I have also created a [PR](https://github.com/w3f/Grants-Program/pull/1444) to change the currency in the original application.

0xCaso commented:
@yyd106 Thanks a lot, now everything is perfect. And no worries, you can close the pull request (it would be necessary if you wanted to change the payment address, but not the ERC20 currency).

I've sent the invoice to the Finance team, so please allow up to 14 days for the payment.

yyd106 commented:
@0xCaso Thanks! Great to work with you

RouvenP commented:
hi @yyd106 we transferred the payment today

yyd106 commented:
Thanks @RouvenP we have received.
